### PR TITLE
build: use webpack-retry-chunk-load-plugin

### PR DESCRIPTION
**Pull request by vm**
Instead of handling retries at the client level in application code by manually wrapping all `import` function usage, we now enable it at the transpilation level.

By using this webpack plugin, we ensure that all code automatically gets retry logic added. We get the added benefit that any npm packages we use also get retry logic added, which was an issue for us with WalletConnect's bundles not being loaded properly.

Implements retry with backoff for 3 attempts. Also reduces the starting retry wait from 1s to .5s.

See that this retries for a blocked URL of a chunk with WalletConnect modal content in it.

Screen Shot 2023-07-06 at 3 25 49 PM

Chunk: https://app.uniswap.org/static/js/4458.0e527482.chunk.js?cache-bust=true&retry-attempt=1
Location of code split: https://github.com/WalletConnect/modal/blob/main/projects/modal/src/client.ts#L21

### CODE REVIEW

File: craco.config.cjs [hunk 1]
```diff
@@ -93,6 +94,10 @@ module.exports = {
       // See https://vanilla-extract.style/documentation/integrations/webpack/#identifiers for docs.
       // See https://github.com/vanilla-extract-css/vanilla-extract/issues/771#issuecomment-1249524366.
       new VanillaExtractPlugin({ identifiers: 'short' }),
+      new RetryChunkLoadPlugin({
+        retryDelay: 2000,
```
Comment (zzmp) [inline craco.config.cjs:98]: This should be a function to add exponential backoff.

File: craco.config.cjs [hunk 2]
```diff
@@ -93,6 +94,10 @@ module.exports = {
       // See https://vanilla-extract.style/documentation/integrations/webpack/#identifiers for docs.
       // See https://github.com/vanilla-extract-css/vanilla-extract/issues/771#issuecomment-1249524366.
       new VanillaExtractPlugin({ identifiers: 'short' }),
+      new RetryChunkLoadPlugin({
```
Comment (zzmp) [inline craco.config.cjs:97]: I think you'll want to include `cacheBust`.

File: craco.config.cjs (hunk 1 above)
Comment (vm) [reply craco.config.cjs:98]: Not possible in their repo, so I'll have to fork if we want that.

File: yarn.lock [hunk 3]
```diff
@@ -16093,10 +16093,10 @@ prettier-linter-helpers@^1.0.0:
   dependencies:
     fast-diff "^1.1.2"
 
-prettier@^2.1.2, prettier@^2.7.1, prettier@^2.8.0:
-  version "2.8.1"
-  resolved "https://registry.yarnpkg.com/prettier/-/prettier-2.8.1.tgz#4e1fd11c34e2421bc1da9aea9bd8127cd0a35efc"
-  integrity sha512-lqGoSJBQNJidqCHE80vqZJHWHRFoNYsSpP9AjFhlhi9ODCJA541svILes/+/1GM3VaL/abZi7cpFzOpdR9UPKg==
+prettier@^2.1.2, prettier@^2.6.2, prettier@^2.7.1, prettier@^2.8.0:
+  version "2.8.8"
```
Comment (zzmp) [inline yarn.lock:16097]: Prettier bumping to 2.8.8 is the reason for all of the "unrelated" lint changes. Can you back this out?
You can still update prettier, but it should be split to a separate PR.

File: craco.config.cjs [hunk 4]
```diff
@@ -93,6 +94,12 @@ module.exports = {
...
+        retryDelay: `function(retryAttempt) {
+          return 2 ** (retryAttempt - 1) * 1000;
```
Comment (zzmp) [inline craco.config.cjs:99]: This first retry will not occur for 1 second. Is this the intended behavior?
I think this is the same as the old behavior - just double-checking.

**Review by zzmp — CHANGES_REQUESTED** [review]
Do you need to add query parameters when retrying? Is it possible (with this plugin) to omit these query parameters?

File: craco.config.cjs (hunk 4 above)
Comment (vm) [reply craco.config.cjs:99]: yeah this is same as existing behavior, but i'm open to changing this around if we want. could do 500ms to start?

**Comment by vm** [comment]
I don't think we need the query params, let me see if I can remove them

File: yarn.lock (hunk 3 above)
Comment (vm) [reply yarn.lock:16097]: I'll make a separate PR 👍

Comment (vm) [reply yarn.lock:16097]: i'll rebase once the other pr is merged

**Review by zzmp — COMMENTED** [review]
pls add back in the cache buster

File: craco.config.cjs [hunk 5]
```diff
@@ -93,6 +94,16 @@ module.exports = {
       // See https://vanilla-extract.style/documentation/integrations/webpack/#identifiers for docs.
       // See https://github.com/vanilla-extract-css/vanilla-extract/issues/771#issuecomment-1249524366.
       new VanillaExtractPlugin({ identifiers: 'short' }),
+      new RetryChunkLoadPlugin({
+        cacheBust: `function() {
+          return 'cache-bust=' + Date.now();
+        }`,
+        // Retries with backoff (500ms, 1000ms, 2000ms).
```
Comment (zzmp) [inline craco.config.cjs:101]:
```suggestion
        // Retries with exponential backoff (500ms, 1000ms, 2000ms).
```

File: craco.config.cjs [hunk 6]
```diff
@@ -93,6 +94,16 @@ module.exports = {
       // See https://vanilla-extract.style/documentation/integrations/webpack/#identifiers for docs.
       // See https://github.com/vanilla-extract-css/vanilla-extract/issues/771#issuecomment-1249524366.
       new VanillaExtractPlugin({ identifiers: 'short' }),
+      new RetryChunkLoadPlugin({
+        cacheBust: `function() {
+          return 'cache-bust=' + Date.now();
```
Comment (zzmp) [inline craco.config.cjs:99]:
```suggestion
          return Date.now();
```

**Review by zzmp — CHANGES_REQUESTED** [review]

File: craco.config.cjs (hunk 6 above)
Comment (vm) [reply craco.config.cjs:99]: you're fine with the URL being `app.uniswap.org/chunk.3.js?253523525`?

**Review by zzmp — APPROVED** [review]

File: craco.config.cjs [hunk 7]
```diff
@@ -93,6 +94,16 @@ module.exports = {
       // See https://vanilla-extract.style/documentation/integrations/webpack/#identifiers for docs.
       // See https://github.com/vanilla-extract-css/vanilla-extract/issues/771#issuecomment-1249524366.
       new VanillaExtractPlugin({ identifiers: 'short' }),
+      new RetryChunkLoadPlugin({
+        cacheBust: `function() {
```
Comment (just-toby) [inline craco.config.cjs:98]: FYI you actually can just define these functions in Javascript above, and call `.toString()` on them

function cacheBust() {
  return 'cache-bust=' + Date.now();
}

function retryDelay(retryAttempt) {
  return 2 ** (retryAttempt - 1) * 500;
}

// Use the functions and convert to strings
new RetryChunkLoadPlugin({
  cacheBust: cacheBust.toString(),
  retryDelay: retryDelay.toString(),
  maxRetries: 3,
});

**Review by just-toby — APPROVED** [review]
